core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -177,6 +178,20 @@ public void run() {
     }
 
     /**
+     * Run stuff synchronously with Queue.lock
+     */
+    private void withLock(Runnable r) throws IOException {
+        try {
+            Queue.withLock(r).get();
+            jenkins.getQueue().settle();
```

Comment:
So you have basically turned the Queue.mutator thread into a lock... I fail to see why we would even bother adding this level of complexity to "simplify" and "remove" the risk of "deadlock"